Fix for removing Kunena via Tools

Ruud68 opened the pull request:
Pull Request for Issue #9923  . 
 
#### Summary of Changes 
set missing database in installer
not tested.

#### Testing Instructions
@rich20 can you try this?

rich20 commented:
The plg_kunena_community extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_comprofiler extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_easysocial extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_gravatar extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_joomla extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_kunena extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_kunena_finder extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_sampledata_kunena extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_finder_kunena extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The plg_quickicon_kunena extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.
The PLG_PRIVACY_KUNENA extension is part of a [package](https://www.rich20.at/fuenf_58321987-78329076534-98091254327-87634563-091423876-8704765-34261558976/administrator/index.php?option=com_installer&view=manage&filter[type]=package&filter[search]=id:262) which does not allow individual extensions to be uninstalled.


<blockquote class="blockquote">
                            <span class="badge bg-secondary">0</span>
                            Database not set in Joomla\CMS\Installer\Installer                        </blockquote>
                                                    <div>
                                <h2 id="caption">Call Stack</h2>

**| Function | Location**
-- | -- | --
1 | () | JROOT/libraries/vendor/joomla/database/src/DatabaseAwareTrait.php:43
2 | Joomla\CMS\Installer\Installer->getDatabase() | JROOT/libraries/src/Installer/Installer.php:2492
3 | Joomla\CMS\Installer\Installer->getAdapter() | JROOT/libraries/src/Installer/Installer.php:904
4 | Joomla\CMS\Installer\Installer->uninstall() | JROOT/libraries/kunena/src/Install/KunenaModelInstall.php:567
5 | Kunena\Forum\Libraries\Install\KunenaModelInstall->uninstallLibrary() | JROOT/libraries/kunena/src/Install/KunenaModelInstall.php:468
6 | Kunena\Forum\Libraries\Install\KunenaModelInstall->uninstall() | JROOT/administrator/components/com_kunena/src/Controller/InstallController.php:77
7 | Kunena\Forum\Administrator\Controller\InstallController->uninstall() | JROOT/libraries/src/MVC/Controller/BaseController.php:730
8 | Joomla\CMS\MVC\Controller\BaseController->execute() | JROOT/libraries/src/Dispatcher/ComponentDispatcher.php:143
9 | Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() | JROOT/libraries/src/Component/ComponentHelper.php:361
10 | Joomla\CMS\Component\ComponentHelper::renderComponent() | JROOT/libraries/src/Application/AdministratorApplication.php:150
11 | Joomla\CMS\Application\AdministratorApplication->dispatch() | JROOT/libraries/src/Application/AdministratorApplication.php:205
12 | Joomla\CMS\Application\AdministratorApplication->doExecute() | JROOT/libraries/src/Application/CMSApplication.php:315
13 | Joomla\CMS\Application\CMSApplication->execute() | JROOT/administrator/includes/app.php:58
14 | require_once() | JROOT/administrator/index.php:32

</div>

rich20 commented:
Addendum: It works with Kunena 6.4.

Ruud68 commented:
We made everything installed via the package dependent on the package. The benefit here is that you cannot break your installation by removing an individual plugin.

What is the use for uninstalling via tools? What value does it add compared to uninstalling via the extension manager?

rich20 commented:
This option completely removes Kunena along with the database tables.

Ruud68 commented:
Yes, but so does the package uninstall, if we configure it to also remove the tables. This is currently 'turned off'. So question is then, if we turn that on again, will the tools uninstall still be needed?
Or is there a reason why not removing the tables on uninstall via extension manager?

rich20 commented:
The Joomla installer also did not remove the “kunena” folder in “/media”. So if problems arose, it was possible to uninstall Kunena and then reinstall it without losing any data.
I cannot judge whether this is actually the purpose, or whether the uninstallation in Tools was only introduced to be able to remove Kunena completely.

Ruud68 commented:
not deleting the kunena tables on uninstall and then installing again worked because everything kunena related was in the tables. Now that we are switching to the Kunena configuration into the component options this will not work as the Kunena component options are stores in #__extensions in the com_kunena record which is deleted on uninstall.
So although all tables are retained on uninstall, installing kunena again will NOT be the same as the kunena configuration is started fresh.

IMO we should following Joomla standards and do a full uninstall (so also including tables and all files = also media), or we do it like AcyMailing does it: on uninstall retain the tables and print on screen the sql commands to remove the tables after unstall from the database.

@xillibit: any thoughts?

xillibit commented:
I think it's more easier for users if the tables are deleted in same time when Kunena is uninstalled instead of doing it by entering a sql query after it. It's a reason vhy we added the uninstall option by using tools

xillibit commented:
Why here are-ther changes about tour other PR : 

[Uninstall via Joomla Extension manager](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/2caaaf114a7a922d2fb160311e7bbcde6ef2c94a)
[2caaaf1](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/2caaaf114a7a922d2fb160311e7bbcde6ef2c94a)
@Ruud68
[Fix for removing Kunena via Tools](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/931d6ba979f172c1b1230d1fd58a9c6eb245ead4)
[931d6ba](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/931d6ba979f172c1b1230d1fd58a9c6eb245ead4)
@Ruud68
[Uninstall via Joomla Extension manager](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/bdcb03203d7355a4024e0866de7108e4f5c90b77)
[bdcb032](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/bdcb03203d7355a4024e0866de7108e4f5c90b77)
@Ruud68
[Merge branch 'uninstallfix' of](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/5bb6226559abfb81c065d9bba580708d413fc4d4) https://github.com/Ruud68/Kunena-Forum […](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/5bb6226559abfb81c065d9bba580708d413fc4d4)
[5](https://github.com/Kunena/Kunena-Forum/pull/9925/commits/5bb6226559abfb81c065d9bba580708d413fc4d4)

?

Ruud68 commented:
I have just removed the uninstall from tools, refactored the package manifest, adjusted the script.php
when you install this version, it will install the package and all plugins / library) except the language package as that is installed seprately.
It will also delete all kunena tables from the database when uninstalling.

@xillibit: the script file used on installation is 'strange' (but works) as this kicks off a complete service provider. I do not know why there wasn't used a regular install script that extends Joomla's InstallerScript as that has a lot of functionality in place (like deleting files and folders) that are now duplicated?

Maybe some legacy code that was never updated / refactored?

Ruud68 commented:
@xillibit  I have just rebased, so the PR you merged is now also in this one (probably not needed to do this, but wanted to avoid merge conflicts)

xillibit commented:
When i had fixed the script to be able to work on Joomla! 6.0, i had keept the legacy code without looking that was code duplicate

xillibit commented:
Just tested, uninstalling Kunena package works like intended

Ruud68 commented:
@xillibit, I can refactor this to an InstallerScript. Not a priority but would make the code more compliant and better readable for others

xillibit commented:
I had chanded it to make working on Joomla! 6.0, before it was an InstallerScript : https://github.com/Kunena/Kunena-Forum/commit/04f3ce351cc81ee0dd42e70535103a813cf8418e

I had followed that : https://manual.joomla.org/docs/next/building-extensions/install-update/installation/install-process/ they said there are two ways the other way is to implement InstallerScriptInterface not InstallerScript

Ruud68 commented:
Thanks  for the explanation. Strange that they advice this in the manual and not the InstallerScript class (from the Core library).
Where the advertised implementation (via the DI) requires you to create all needed (and required) functions, the InstallerScript already has these, including useful methods for e.g. removing files and folders (where you now have created your own code for which also needs to be maintained), getting parameters etc.

Ruud68 commented:
While deep diving into the kunena media package, I see that the scripts we are using for the plugins etc. are used on 'legacy' support:
```
            @trigger_error(
                'Legacy installer files are deprecated and will be removed in 6.0. Use a service provider instead.',
                E_USER_DEPRECATED
            );
```

But this still remains to be seen. In J6 there is a InstallerScriptTrait that has all the 'old' InstallerScript functions (like delete folders etc.), but in Joom 5.4 there is no such trait.

So either they remove it (which IMO is very late in the process) and then also add the trait on J5.4 for compatibilty, or they remove the functionality in Joomla 7.0

Let's wait for the Joomla 6 RC and 5.4 RC version to see what has been removed for real

xillibit commented:
It's for today the release of Joomla 6 RC and 5.4 RC, i'am following the changes in these versions but seems only bugfixes others changes are now directed to J!6.1

Ruud68 commented:
> It's for today the release of Joomla 6 RC and 5.4 RC, i'am following the changes in these versions but seems only bugfixes others changes are now directed to J!6.1

https://github.com/joomla/joomla-cms/issues/46205
